src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs

```diff
@@ -89,54 +89,84 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
         /// <inheritdoc />
         public async Task<string> SetupAsync(SetupContext context)
         {
+            ShellSettings shellSettings = null;
             var initialState = context.ShellSettings.State;
             try
             {
-                var executionId = await SetupInternalAsync(context);
+                // Set the shell state to 'Initializing' to guarantee that any subsequent HTTP requests
+                // will result in a 'Service Unavailable' response while the tenant setup is in progress.
+                context.ShellSettings.AsInitializing();
 
-                if (context.Errors.Any())
+                (var executionId, shellSettings) = await SetupInternalAsync(context);
+
+                if (context.Errors.Count > 0)
                 {
                     context.ShellSettings.State = initialState;
-                    await _shellHost.ReloadShellContextAsync(context.ShellSettings, eventSource: false);
+                    await _shellHost.ReloadShellContextAsync(shellSettings, eventSource: false);
+
+                    var shellScope = await _shellHost.GetScopeAsync(shellSettings);
+                    await shellScope.UsingAsync(async scope =>
+                    {
+                        var tenandSetupHandlers = scope.ServiceProvider.GetServices<ITenantSetupHandler>();
+
+                        await tenandSetupHandlers.InvokeAsync((handler) => handler.FailedAsync(context), _logger);
+                    });
+
+                    return null;
                 }
 
+                // At this point, we know that the setup was successful.
+                // Set the shell state to 'Running' state.
+                await _shellHost.UpdateShellSettingsAsync(shellSettings.AsRunning());
+
+                var runningShellScope = await _shellHost.GetScopeAsync(shellSettings);
+                await runningShellScope.UsingAsync(async scope =>
+                {
+                    var tenandSetupHandlers = scope.ServiceProvider.GetServices<ITenantSetupHandler>();
+
+                    await tenandSetupHandlers.InvokeAsync((handler) => handler.SucceededAsync(), _logger);
```

Comment:
That's okay for the succeeded event, here or as we did previously.

If we only have this new event, we could just add this as I suggested but by not using a background job.

If we keep the `Failed` event and if it should be managed differently (not by the related tenant itself), maybe better to do the same for the `Suceeded` event too.

Reply:
I moved the `FailedAsync` call just after we call `SetupAsync()` so we can report the error in the same scope. Let me know if you see other issues here.